### ISSUE TITLE
Google Pixel 3 XL - SELinux prevents "adb shell" from running root commands

### ISSUE THREAD

**Issue by adpoliak** [topjohnwu/Magisk#2960]
Magisk Canary: 97b72a59 (20419)
Magisk Manager: 9348c5ba (289)

Google Pixel 3 XL (Crosshatch) running `10.0.0 (QQ3A.200605.001, Jun 2020)` images + Magisk + ElementalX-P3-2.16 Kernel

Unsure if this is a Magisk and/or ElementalX issue, but running `adb shell 'su -c "COMMAND"'` was failing on my device, while running `adb shell` then typing in `su -c "COMMAND"` would work.

I tracked it down to a SELinux issue via the use of `setenforce 0`, seeing the previously-failing command succeed, then running `setenforce 1`.
The following was logged:
```
avc: denied { read write } for comm="su" path="socket:[22204983]" dev="sockfs" ino=22204983 scontext=u:r:magisk_client:s0 tcontext=u:r:adbd:s0 tclass=unix_stream_socket permissive=1
avc: denied { read write } for comm="su" path="socket:[22204983]" dev="sockfs" ino=22204983 scontext=u:r:magisk_client:s0 tcontext=u:r:adbd:s0 tclass=unix_stream_socket permissive=1
avc: denied { read } for comm="su" name="bin" dev="dm-1" ino=21 scontext=u:r:magisk_client:s0 tcontext=u:object_r:vendor_file:s0 tclass=dir permissive=1
avc: denied { read } for comm="su" name="bin" dev="dm-1" ino=21 scontext=u:r:magisk_client:s0 tcontext=u:object_r:vendor_file:s0 tclass=dir permissive=1
avc: denied { read } for comm="su" name="tmp" dev="dm-3" ino=65 scontext=u:r:magisk_client:s0 tcontext=u:object_r:shell_data_file:s0 tclass=dir permissive=1
avc: denied { read } for comm="su" name="tmp" dev="dm-3" ino=65 scontext=u:r:magisk_client:s0 tcontext=u:object_r:shell_data_file:s0 tclass=dir permissive=1
avc: denied { getattr } for comm="su" path="/data" dev="dm-3" ino=3 scontext=u:r:magisk_client:s0 tcontext=u:object_r:system_data_file:s0 tclass=dir permissive=1
avc: denied { getattr } for comm="su" path="/data" dev="dm-3" ino=3 scontext=u:r:magisk_client:s0 tcontext=u:object_r:system_data_file:s0 tclass=dir permissive=1
avc: denied { getattr } for comm="su" path="/data/local/tmp" dev="dm-3" ino=65 scontext=u:r:magisk_client:s0 tcontext=u:object_r:shell_data_file:s0 tclass=dir permissive=1
avc: denied { getattr } for comm="su" path="/data/local/tmp" dev="dm-3" ino=65 scontext=u:r:magisk_client:s0 tcontext=u:object_r:shell_data_file:s0 tclass=dir permissive=1
avc: denied { getattr } for comm="su" path="/postinstall" dev="dm-0" ino=48 scontext=u:r:magisk_client:s0 tcontext=u:object_r:postinstall_mnt_dir:s0 tclass=dir permissive=1
avc: denied { getattr } for comm="su" path="/postinstall" dev="dm-0" ino=48 scontext=u:r:magisk_client:s0 tcontext=u:object_r:postinstall_mnt_dir:s0 tclass=dir permissive=1
avc: denied { read } for comm="su" name="bin" dev="dm-1" ino=21 scontext=u:r:magisk_client:s0 tcontext=u:object_r:vendor_file:s0 tclass=dir permissive=1
avc: denied { read } for comm="su" name="bin" dev="dm-1" ino=21 scontext=u:r:magisk_client:s0 tcontext=u:object_r:vendor_file:s0 tclass=dir permissive=1
avc: denied { read } for comm="su" name="tmp" dev="dm-3" ino=65 scontext=u:r:magisk_client:s0 tcontext=u:object_r:shell_data_file:s0 tclass=dir permissive=1
avc: denied { read } for comm="su" name="tmp" dev="dm-3" ino=65 scontext=u:r:magisk_client:s0 tcontext=u:object_r:shell_data_file:s0 tclass=dir permissive=1
avc: denied { getattr } for comm="su" path="/data" dev="dm-3" ino=3 scontext=u:r:magisk_client:s0 tcontext=u:object_r:system_data_file:s0 tclass=dir permissive=1
avc: denied { getattr } for comm="su" path="/data" dev="dm-3" ino=3 scontext=u:r:magisk_client:s0 tcontext=u:object_r:system_data_file:s0 tclass=dir permissive=1
```

After patching the SELinux policy with the following, the `adb shell 'su -c "COMMAND"'` construct worked:
```
magiskpolicy --live \
'allow magisk_client adbd unix_stream_socket {read write ioctl}' \
'allow magisk magisk_file sock_file {ioctl}' \
'allow magisk_client adbd process {sigchld}' \
'allow magisk_file sockfs filesystem {associate}' \
'allow magisk_client vendor_file dir {read}' \
'allow magisk_client shell_data_file dir {getattr read}' \
'allow magisk_client system_data_file dir {getattr}' \
'allow magisk_client postinstall_mnt_dir dir {getattr}' 
```

I'm hesitant to submit a PR containing modifications to native/jni/magiskpolicy/rules.cpp if my device is the only one affected by this.

**Comment by AndDiSa** [comment]
I can confirm the issue on a magisk-patched emulator device. After applying the additional rules the adb shell command is working again.

**Comment by UltraBlackLinux** [comment]
For me it worked to type su and go into magisk to grant root access

**Comment by AndDiSa** [comment]
Yes, manually you can do so, but you cannot do scripting, i.e.

adb shell su 0 -c "ls"

will not work.

**Comment by topjohnwu** [comment]
Should be fixed in 4dacffd7a195ef0ae0ec23999185407237e7805f
